### PR TITLE
Install-PSResource -RequiredResource with empty inner hashtable value bugfix

### LINKED ISSUE #1411: Anything with RequiredResource fails silently
### Prerequisites

- [X] Write a descriptive title.
- [X] Make sure you are able to repro it on the [latest released version](https://www.powershellgallery.com/packages/PSResourceGet)
- [X] Search the existing issues.

### Steps to reproduce

I thought I was supposed to be able to run commands like this:

```PowerShell
Install-PSResource -RequiredResourceFile .\RequiredModules.psd1
Install-PSResource -RequiredResource @{ Pester = "[5.3.0,6.0)" }
```

But this does not do anything, and when I run it with `-Verbose` it does not show something like this:
`Installed package Pester 5.5.0 is already available.`

When I try something ridiculous like this:

```PowerShell
Install-PSResource -RequiredResource @{
    ModuleName = "There is No Module With This Name"
    ModuleVersion = "A.B.C"
}
```

The same _nothing_ happens.

### Expected behavior

```console
Missing modules should be installed
```

### Actual behavior

```console
Nothing
```

### Error details

_No response_

### Environment data

```powershell
ModuleType Version    PreRelease Name                                ExportedCommands
---------- -------    ---------- ----                                ----------------
Binary     0.9.0      rc1        Microsoft.PowerShell.PSResourceGet  {Find-PSResource, Get-InstalledPSResource, Get-PSResourceRepository, Get-PSScriptFileInfo…}


Name                           Value
----                           -----
PSVersion                      7.4.0-preview.5
PSEdition                      Core
GitCommitId                    7.4.0-preview.5
OS                             Microsoft Windows 10.0.19042
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

### Visuals

_No response_

### PR DESCRIPTION
**Pull request by anamnavi**
# PR Summary
For examples of valid input for Install-PSResource's `-RequiredResource` cmdlet please reference the docs [here](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.psresourceget/install-psresource?view=powershellget-3.x#example-5)

Additionally parameters can be supplied via parameter splatting which is supported in powershell.

## PR Context
Resolves #1411 

## PR Checklist

- [ ] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [ ] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ ] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ ] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [ ] None
    - **OR**
    - [ ] [Documentation needed]()
        - [ ] Issue filed: 
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [ ] [Documentation needed]()
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [ ] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary.

### CODE REVIEW

File: src/code/InstallPSResource.cs [hunk 1]
```diff
@@ -426,50 +426,63 @@ protected override void ProcessRecord()
...
-
-                // Format should now be a hashtable, whether the original input format was json or hashtable
-                if (!(pkgParamInfo is Hashtable pkgInstallInfo))
+                // Set package name which will be key for the inner hashtable and is present for all scenarios,
+                // inclding the scenario where only package name is specified
+                // i.e Install-PSResource -RequiredResource @ { MyPackage = @{} }
+                string pkgName = entry.Key.ToString();
+                if (!(entry.Value is Hashtable pkgInstallInfo))
                 {
+                    var requiredReourceHashtableInputFormatError = new ErrorRecord(
```
Comment (alerickson) [inline src/code/InstallPSResource.cs]:
```suggestion
                    var requiredResourceHashtableInputFormatError = new ErrorRecord(
```

File: src/code/InstallPSResource.cs [hunk 2]
```diff
@@ -426,50 +426,63 @@ protected override void ProcessRecord()
...
+                        new ArgumentException($"The RequiredResource input with name {pkgName} does not have a valid value, the value must be a hashtable."),
+                        "RequiredReourceHashtableInputFormatError",
```
Comment (alerickson) [inline src/code/InstallPSResource.cs]:
```suggestion
                        "RequiredResourceHashtableInputFormatError", 
```

File: src/code/InstallPSResource.cs [hunk 3]
```diff
@@ -426,50 +426,63 @@ protected override void ProcessRecord()
...
+                string pkgName = entry.Key.ToString();
+                if (!(entry.Value is Hashtable pkgInstallInfo))
                 {
+                    var requiredReourceHashtableInputFormatError = new ErrorRecord(
+                        new ArgumentException($"The RequiredResource input with name {pkgName} does not have a valid value, the value must be a hashtable."),
+                        "RequiredReourceHashtableInputFormatError", 
+                        ErrorCategory.InvalidArgument,
+                        this);
+
+                    WriteError(requiredReourceHashtableInputFormatError);
```
Comment (alerickson) [inline src/code/InstallPSResource.cs]:
```suggestion
                    WriteError(requiredResourceHashtableInputFormatError);
```

File: src/code/InstallPSResource.cs [hunk 4]
```diff
@@ -426,50 +426,63 @@ protected override void ProcessRecord()
...
-                if (!(pkgParamInfo is Hashtable pkgInstallInfo))
+                // Set package name which will be key for the inner hashtable and is present for all scenarios,
+                // inclding the scenario where only package name is specified
+                // i.e Install-PSResource -RequiredResource @ { MyPackage = @{} }
+                string pkgName = entry.Key.ToString();
+                string pkgVersion = String.Empty;
+                if (!(entry.Value is Hashtable pkgInstallInfo))
                 {
+                    var requiredResourceHashtableInputFormatError = new ErrorRecord(
+                        new ArgumentException($"The RequiredResource input with name {pkgName} does not have a valid value, the value must be a hashtable."),
```
Comment (anamnavi) [inline src/code/InstallPSResource.cs]:
```suggestion
                        new ArgumentException($"The RequiredResource input with name '{pkgName}' does not have a valid value, the value must be a hashtable."),
```

File: src/code/InstallPSResource.cs [hunk 5]
```diff
@@ -426,50 +426,63 @@ protected override void ProcessRecord()
...
+            foreach (DictionaryEntry entry in reqResourceHash)
             {
-                var pkgParamInfo = reqResourceHash[pkgName];
+                InstallPkgParams pkgParams = new InstallPkgParams();
+                PSCredential pkgCredential = Credential;
 
-                // Format should now be a hashtable, whether the original input format was json or hashtable
-                if (!(pkgParamInfo is Hashtable pkgInstallInfo))
+                // Set package name which will be key for the inner hashtable and is present for all scenarios,
+                // inclding the scenario where only package name is specified
```
Comment (anamnavi) [inline src/code/InstallPSResource.cs]:
```suggestion
                // including the scenario where only package name is specified
```

File: src/code/InstallPSResource.cs [hunk 6]
```diff
@@ -426,50 +426,63 @@ protected override void ProcessRecord()
...
-            foreach (string pkgName in pkgNames)
+            foreach (DictionaryEntry entry in reqResourceHash)
             {
-                var pkgParamInfo = reqResourceHash[pkgName];
+                InstallPkgParams pkgParams = new InstallPkgParams();
+                PSCredential pkgCredential = Credential;
 
-                // Format should now be a hashtable, whether the original input format was json or hashtable
-                if (!(pkgParamInfo is Hashtable pkgInstallInfo))
+                // Set package name which will be key for the inner hashtable and is present for all scenarios,
```
Comment (anamnavi) [inline src/code/InstallPSResource.cs]:
```suggestion
                // The package name which will be the key for the inner hashtable and is present for all scenarios,
```

File: src/code/InstallPSResource.cs [hunk 7]
```diff
@@ -426,50 +426,63 @@ protected override void ProcessRecord()
...
-            foreach (string pkgName in pkgNames)
+            foreach (DictionaryEntry entry in reqResourceHash)
             {
-                var pkgParamInfo = reqResourceHash[pkgName];
+                InstallPkgParams pkgParams = new InstallPkgParams();
+                PSCredential pkgCredential = Credential;
 
-                // Format should now be a hashtable, whether the original input format was json or hashtable
-                if (!(pkgParamInfo is Hashtable pkgInstallInfo))
+                // The package name which will be the key for the inner hashtable and is present for all scenarios,
```
Comment (anamnavi) [inline src/code/InstallPSResource.cs]:
```suggestion
                // The package name will be the key for the inner hashtable and is present for all scenarios,
```

File: src/code/InstallPSResource.cs [hunk 8]
```diff
@@ -426,50 +426,63 @@ protected override void ProcessRecord()
...
-                    {
-                        WriteVerbose("Credential specified for required resource");
-                        pkgCredential = pkgInstallInfo[paramName] as PSCredential;
-                    }
+                    var pkgParamNames = pkgInstallInfo.Keys;
 
-                    pkgParams.SetProperty(paramName, pkgInstallInfo[paramName] as string, out ErrorRecord IncorrectVersionFormat);
+                    foreach (string paramName in pkgParamNames)
+                    {
+                        if (string.Equals(paramName, "credential", StringComparison.InvariantCultureIgnoreCase))
```
Comment (adityapatwardhan) [inline src/code/InstallPSResource.cs]: Declare "credential" as a const outside the loop. Having a string inside the loop causes allocations every iteration.

**Review by adityapatwardhan — APPROVED** [review]

File: src/code/InstallPSResource.cs [hunk 9]
```diff
@@ -426,50 +426,63 @@ protected override void ProcessRecord()
...
-                string pkgVersion = pkgInstallInfo["version"] == null ? String.Empty : pkgInstallInfo["version"].ToString();
+                    pkgVersion = pkgInstallInfo["version"] == null ? String.Empty : pkgInstallInfo["version"].ToString();
+                }
 
                 ProcessInstallHelper(
                     pkgNames: new string[] { pkgName },
                     pkgVersion: pkgVersion,
                     pkgPrerelease: pkgParams.Prerelease,
-                    pkgRepository: new string[] { pkgParams.Repository },
+                    pkgRepository: pkgParams.Repository != null ? new string[] { pkgParams.Repository } : new string[]{},
```
Comment (alerickson) [inline src/code/InstallPSResource.cs:499]: Do we need to add a null check for all the parameters?  Do you know what happens if, for example, a pkgVersion, isn't passed in by the user?

**Review by alerickson — APPROVED** [review]

File: src/code/InstallPSResource.cs (hunk 9 above)
Comment (anamnavi) [reply src/code/InstallPSResource.cs:499]: pkgVersion will be an empty string if not provided via the RequiredResource hashtable entry. I added a check for pkgName to not be empty or whitespace.

**Pull request merged**